### PR TITLE
[PATRON2020-15] Mqtt gateway library

### PR DESCRIPTION
**Pull request by Deelay1403**
Library supports sending and receiving information via MQTT
main.ino contain 2 sample functions, which won't be available in final version of lib

### CODE REVIEW

**Comment by Deelay1403** [comment]
@szymoneiro setReturnFunct(String) sets a pointer on the function that callback should be returned.

When a message arrives at the device, it's compressed as a payload array. The callback function waits for the whole payload and then returns it as a string.

f.eq. if you prepare function like this:
void print_input(String s){ Serial.println(s); }

and then:

mgtt.setReturnFunct(print_input);

Every time if the device receives the message, the library will execute your function and parse the string to it.

**Comment by Deelay1403** [comment]
"I couldn't find a function which sends any kind of message at connection and disconnection"
I don't think that's necessary. The library checks the status of the connection and reconnects if necessary. With knowledge about the connection, you couldn't do anything.

File: MQTT/main.ino [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+void setup() {
+  /* Initialize serial and wait for port to open: */
+  Serial.begin(115200);
+  while (!Serial) {
+    ; /* wait for serial port to connect. Needed for native USB port only */
+  }
+  pinMode(LED, OUTPUT);
+  digitalWrite(LED, HIGH);
+
+  // attempt to connect to Wifi network:
```
Comment (kuba8686) [inline MQTT/main.ino]: comment should start with a capital letter due to formating rules

File: MQTT/MQTT.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#include "MQTT.h"
+
+void MQTT::reconnect() {
+  // Loop until we're reconnected
+  while (!client.connected()) {
```
Comment (Siwy160) [inline MQTT/MQTT.cpp]: 4 spaces as intends please.

File: MQTT/MQTT.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+#include <ESP8266WiFi.h>        // Include the Wi-Fi library
+#include <Arduino.h>
+
+#define DEBUG 1
+
+// Public broker
+// const char broker[] = "broker.hivemq.com";
+// int        port     = 1883;
+// const char topic[]  = "intive/embedded";
+// const char outTopic[] = "intive/embedded_o";
```
Comment (Siwy160) [inline MQTT/MQTT.h]: We are using `/* comment */` style comments.

File: MQTT/MQTT.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+// Our broker
+const char broker[] = "pszczodrowski.pl";
+int        port     = 1883;
+const char topic[]  = "intive/embedded";
+const char outTopic[] = "intive/embedded_o";
```
Comment (Siwy160) [inline MQTT/MQTT.h]: I see that you tried to lineup declarations nicely. (and you failed btw)

Just make it:
```c
const char broker[] = "pszczodrowski.pl";
int port = 1883;
const char topic[] = "intive/embedded";
const char outTopic[] = "intive/embedded_o";
```

File: MQTT/MQTT.cpp (hunk 2 above)
Comment (Deelay1403) [reply MQTT/MQTT.cpp]: Solved

File: MQTT/MQTT.h (hunk 3 above)
Comment (Deelay1403) [reply MQTT/MQTT.h]: Solved

File: MQTT/MQTT.h (hunk 4 above)
Comment (Deelay1403) [reply MQTT/MQTT.h]: Solved

File: MQTT/main.ino (hunk 1 above)
Comment (Deelay1403) [reply MQTT/main.ino]: Solved

File: MQTT/MQTT.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+int port = 1883; 
+const char topic[] = "intive/embedded";
+const char outTopic[] = "intive/embedded_o"; 
+*/
+
+/* Our broker */
+const char broker[] = "pszczodrowski.pl";
+int port = 1883;
+const char topic[] = "intive/embedded";
+const char outTopic[] = "intive/embedded_o";
```
[the remark this answers is not in the document]
Comment (szymoneiro) [reply MQTT/MQTT.h]: Well, I had the same issue but found a solution, not gonna tell you Patryk then :D

[the remark this answers is not in the document]
Comment (Deelay1403) [reply MQTT/MQTT.h]: I see my mistake. This variable should be const.

File: MQTT/main.ino [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  2) Change port value from 1883 to 8883.
+  3) Change broker value to a server with a known SSL/TLS root certificate 
+    flashed in the WiFi module. */
+MQTT mqtt;
+
+/*  Sample input parser. Only for presentation.  */
+void led_plain(String s){
+    if(s[0]=='0') digitalWrite(LED,HIGH);
+    else if(s[0]=='1') digitalWrite(LED,LOW);
+}
```
[the remark this answers is not in the document]
Comment (Deelay1403) [reply MQTT/main.ino]: The whole .ino file was only for presentation purposes. I made this function only for testing and I wasn't sure that I'll prepare JSON parsing in time.

Should I remove tracking from .ino file?

File: MQTT/MQTT.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
+#include "MQTT.h"
+
+void MQTT::reconnect() {
+    /* Loop until we're reconnected */
+    while (!client.connected()) {
+        #if DEBUG==1
```
[the remark this answers is not in the document]
Comment (Deelay1403) [reply MQTT/MQTT.cpp]: That's a very good point.
In that case, I think we should rethink about how we will add debug options to our code.

File: MQTT/main.ino [hunk 8]
```diff
@@ -0,0 +1,63 @@
+/*
```
[the remark this answers is not in the document]
Comment (Deelay1403) [reply MQTT/main.ino]: Done

File: MQTT/main.ino [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    else if(s[0]=='1') digitalWrite(LED,LOW);
+}
+/*  Sample json parser. Only for presentation.  */
+void led_json(String s){
+    deserializeJson(JSONdoc, s);
+    if(JSONdoc["command"].as<String>().equals("LED"))
+        if(JSONdoc["state"].as<int>() == 0)
+            digitalWrite(LED, HIGH);
+        else if(JSONdoc["state"].as<int>() == 1)
+            digitalWrite(LED, LOW);
```
[the remark this answers is not in the document]
Comment (Deelay1403) [reply MQTT/main.ino]: Done

File: MQTT/MQTT.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (cichy1024) [reply MQTT/MQTT.cpp]: Debug is OK - the only thing is there should be no indentation for preprocessor definitions but only for our code, e.g:
```
void foo(void) {
    while (true) {
#if DEBUG==1
        /* Do debug stuff */
#endif
        /* Do something */
    }
}
```

File: MQTT/MQTT.cpp [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            #if DEBUG==1
+            Serial.print("failed, rc=");
+            Serial.print(client.state());
+            Serial.println(" try again in 5 seconds");
+            #endif
+          /* Wait 5 seconds before retrying */
+            delay(5000);
+      }
+    }
+  }
```
Comment (Siwy160) [inline MQTT/MQTT.cpp]: Intends are still wrong ;)

File: MQTT/MQTT.ino [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    // }
+    pinMode(LED, OUTPUT);
+    digitalWrite(LED, HIGH);
+
+    /* Attempt to connect to Wifi network: */
+    Serial.print("Attempting to connect to WPA SSID: ");
+    Serial.println(ssid);
+    WiFi.begin(ssid, pass);
+    while (WiFi.status() != WL_CONNECTED) {
+      /* Failed, retry */
```
Comment (Siwy160) [inline MQTT/MQTT.ino]: Here as well. There should be 4 spaces indentation.

File: MQTT/MQTT.ino [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    else if(s[0]=='1') digitalWrite(LED,LOW);
+}
+/*  Sample json parser. Only for presentation.  */
+void led_json(String s){
+    deserializeJson(JSONdoc, s);
+    if(JSONdoc["command"].as<String>().equals("LED"))
+        if(JSONdoc["state"].as<int>())
+            digitalWrite(LED, LOW);
+        else if(JSONdoc["state"].as<int>())
+            digitalWrite(LED, HIGH);
```
Comment (cichy1024) [inline MQTT/MQTT.ino]: This condition will never be executed.

**Pull request merged**